### PR TITLE
WIP: Initial Implementation of vi mode

### PR DESCRIPTION
**Pull request by deepakchethan**
The current implementation supports:
**Insertion**: i, a, I, A, r and p
**Movement**: h, l, 0, ^, $, w, b, e, W, B, E
**Deletion**:  d, dw, db, dB, dW, dh, dl, d$, d^, d0, D
c, cw, cb, cB, cW, ch, cl, c$, c^, c0 and x
ciw/diw, cib/dib
**others**: Arrow keys

It also fixes two other issues: 
* #14470: Close popup widget on escape
* #6321: Wrong handling of escape

Please let me know if any other command is to be implemented.

The numbering logic is not implemented yet, I am working on it please don't merge

### CODE REVIEW

**Comment by XVilka** [comment]
See 

E.g. missing common are:
- `^` — Move to start of line
- `$` — Move to end of line
- `e` — Move to the end of the next word
- `diw`/`ciw` - Change the word

**Comment by XVilka** [comment]
Please also document it in the book once finished: https://github.com/radare/radare2book/issues/208

**Comment by radare** [comment]
you cant use rconfig from rcons. add a field in the RConsInstance struct

```
/home/travis/build/radare/radare2/build/../libr/cons/dietline.c:1772: undefined reference to `r_config_get_i'
```

File: libr/core/cconfig.c [hunk 1]
```diff
@@ -2955,6 +2955,7 @@ R_API int r_core_config_init(RCore *core) {
 	SETPREF ("asm.noisy", "true", "Show comments considered noisy but possibly useful");
 	SETPREF ("asm.offset", "true", "Show offsets at disassembly");
 	SETPREF ("scr.square", "true", "Use square pixels or not");
+	SETPREF ("scr.vi", "false", "Use vi mode");
```
Comment (radare) [inline libr/core/cconfig.c]: "Use vi mode for input prompt"

File: libr/cons/dietline.c [hunk 2]
```diff
@@ -1196,12 +1250,271 @@ R_API const char *r_line_readline_cb_win(RLineReadCallback cb, void *user) {
...
+				}
+			}
+			if (i >= I.buffer.length) {
+				I.buffer.index = I.buffer.length;
+			}
+			__print_prompt();
+		} break;
+		case 'w': 
+			while (rep--) {
+			for (i = I.buffer.index + 1; i < I.buffer.length; i++) {
```
Comment (radare) [inline libr/cons/dietline.c]: wrong indent

File: libr/cons/dietline.c [hunk 3]
```diff
@@ -1196,12 +1250,271 @@ R_API const char *r_line_readline_cb_win(RLineReadCallback cb, void *user) {
...
+					break;
+				}
+			}
+			if (i < 0) {
+				I.buffer.index = 0;
+			}
+			__print_prompt();
+		} break;
+		case 'W': 
+			while (rep--) {
```
Comment (radare) [inline libr/cons/dietline.c]: wrong indent

File: libr/cons/dietline.c [hunk 4]
```diff
@@ -1196,12 +1250,271 @@ R_API const char *r_line_readline_cb_win(RLineReadCallback cb, void *user) {
...
+		
+#else
+		I.buffer.index = I.buffer.index < I.buffer.length
+			? I.buffer.index + 1
+			: I.buffer.length;
+#endif
+			__print_prompt();
+		} break;
+		case 'B': 
+			while (rep--) {
```
Comment (radare) [inline libr/cons/dietline.c]: looks like a bit spaguetti and bad indent

File: libr/cons/cons.c [hunk 5]
```diff
@@ -635,6 +635,14 @@ R_API int r_cons_eof() {
 	return feof (I.fdin);
 }
 
+R_API void r_cons_go_forward (int x) {
+	printf ("\x1b[%dC", x);
```
Comment (radare) [inline libr/cons/cons.c]: uhm.. not sure how is this  going  to be used

**Comment by XVilka** [comment]
@deepakchethan don't forget to add them, once implemented in https://github.com/radare/radare2book/blob/master/basic_commands/dietline.md book chapter.

File: libr/cons/cons.c (hunk 5 above)
Comment (deepakchethan) [reply libr/cons/cons.c]: Yea, it is useless. I was trying out a different approach initially and forgot to undo it. I will remove it.

File: libr/cons/dietline.c [hunk 6]
```diff
@@ -1196,12 +1256,364 @@ R_API const char *r_line_readline_cb_win(RLineReadCallback cb, void *user) {
...
+				}
+				if (i >= I.buffer.length) {
+					I.buffer.index = I.buffer.length - 1;
+				}
+				__print_prompt();
+		} break;
+		case 'e':
+			while (rep--) {
+				for (i = I.buffer.index + 1; i < I.buffer.length; i++) {
+					if ((!is_word_break_char (I.buffer.data[i], MINOR_BREAK) && is_word_break_char (I.buffer.data[i+1], MINOR_BREAK))
```
Comment (XVilka) [inline libr/cons/dietline.c]: What about separating these into small, maybe inline functions?

File: libr/core/cconfig.c (hunk 1 above)
Comment (XVilka) [reply libr/core/cconfig.c]: What about using `scr.prompt.vi` instead? There is one for autocompletion widget for example...

File: libr/cons/dietline.c (hunk 4 above)
Comment (deepakchethan) [reply libr/cons/dietline.c]: fixed indent.

**Comment by deepakchethan** [comment]
> you cant use rconfig from rcons. add a field in the RConsInstance struct
> 
> ```
> /home/travis/build/radare/radare2/build/../libr/cons/dietline.c:1772: undefined reference to `r_config_get_i'
> ```

Since this is a dietline config, isn't it better to add it to RLine instance ?

File: libr/core/cconfig.c [hunk 7]
```diff
@@ -2955,6 +2962,7 @@ R_API int r_core_config_init(RCore *core) {
 	SETPREF ("asm.noisy", "true", "Show comments considered noisy but possibly useful");
 	SETPREF ("asm.offset", "true", "Show offsets at disassembly");
 	SETPREF ("scr.square", "true", "Use square pixels or not");
+	SETCB ("scr.vi", "false", &cb_scr_vi, "Use vi mode for input prompt");
```
Comment (XVilka) [inline libr/core/cconfig.c]: please change to the `scr.prompt.vi`

Comment (deepakchethan) [reply libr/core/cconfig.c]: changed it :)

File: libr/cons/dietline.c (hunk 6 above)
Comment (deepakchethan) [reply libr/cons/dietline.c]: Updated the code :)

**Comment by XVilka** [comment]
@deepakchethan please split this long `if` lines into multiple, like:
```c
if ((!is_word_break_char (I.buffer.data[i], MINOR_BREAK)
    && is_word_break_char (I.buffer.data[i - 1], MINOR_BREAK))
    || (is_word_break_char (I.buffer.data[i], MINOR_BREAK)
    && !is_word_break_char (I.buffer.data[i], MAJOR_BREAK))) {
```

File: libr/cons/dietline.c [hunk 8]
```diff
@@ -1196,12 +1255,385 @@ R_API const char *r_line_readline_cb_win(RLineReadCallback cb, void *user) {
...
+
+static void __move_cursor_left () {
+	int utflen;
+#if USE_UTF8
+	{
+		char *s = I.buffer.data + I.buffer.index - 1;
+		utflen = 1;
+		while (s > I.buffer.data && (*s & 0xc0) == 0x80) {
+			utflen++;
+				s--;
```
Comment (XVilka) [inline libr/cons/dietline.c]: Also here - please fix the indent.

File: libr/cons/dietline.c [hunk 9]
```diff
@@ -1196,12 +1255,394 @@ R_API const char *r_line_readline_cb_win(RLineReadCallback cb, void *user) {
...
+	fflush (stdout);
+}
+
+static void __move_cursor_right () {
+	int utflen;
+#if USE_UTF8
+	{
+		char *s = I.buffer.data + I.buffer.index + 1;
+		utflen = 1;
+			while ((*s & 0xc0) == 0x80) {
```
Comment (XVilka) [inline libr/cons/dietline.c]: Indentation still broken here.

File: libr/cons/dietline.c [hunk 10]
```diff
@@ -1472,9 +1916,8 @@ R_API const char *r_line_readline_cb(RLineReadCallback cb, void *user) {
 				break;
 			case 'B':
 			case 'b':
-				// previous word
 				for (i = I.buffer.index - 2; i >= 0; i--) {
-					if (is_word_break_char (I.buffer.data[i]) && !is_word_break_char (I.buffer.data[i + 1])) {
+					if (is_word_break_char (I.buffer.data[i], MINOR_BREAK) && !is_word_break_char (I.buffer.data[i + 1], MINOR_BREAK)) {
```
Comment (XVilka) [inline libr/cons/dietline.c]: What about splitting these long ifs too?

File: libr/cons/dietline.c (hunk 9 above)
Comment (deepakchethan) [reply libr/cons/dietline.c]: I apologize for that, it seemed fine in vscode :/

**Pull request merged**